_posts/2020-10-01-torch-network-from-scratch/torch_network_from_scratch.Rmd

```diff
@@ -0,0 +1,1076 @@
+
+## A simple neural network using `torch` tensors
+
+Our task, which we'll approach in a low-level way today but considerably simplify in upcoming installments, consists of regressing a single target datum based on three input variables.
+
+We directly use `torch` to simulate some data.
+
+#### Toy data
+
+```{r}
+library(torch)
+
+# input dimensionality (number of input features)
+d_in <- 3
+# output dimensionality (number of predicted features)
+d_out <- 1
+# number of observations in training set
+n <- 100
+
+
+# create random data
+# input
+x <- torch_randn(n, d_in)
+# target
+y <- x[, 1, drop = FALSE] * 0.2 -
+  x[, 2, drop = FALSE] * 1.3 -
+  x[, 3, drop = FALSE] * 0.5 +
+  torch_randn(n, 1)
+```
+
+Next, we need to initialize the network's weights. We'll have one hidden layer, with `32` units. The output layer's size, being determined by the task, is equal to `1`.
+
+#### Initialize weights
+
+```{r}
+# dimensionality of hidden layer
+d_hidden <- 32
+
+# weights connecting input to hidden layer
+w1 <- torch_randn(d_in, d_hidden)
+# weights connecting hidden to output layer
+w2 <- torch_randn(d_hidden, d_out)
+
+# hidden layer bias
+b1 <- torch_zeros(1, d_hidden)
+# output layer bias
+b2 <- torch_zeros(1, d_out)
+```
+
+Now for the training loop proper. The training loop really here *is* the network.
+
+#### Training loop
+
+In each iteration ("epoch"), the training loop does four things:
+
+-   run through the network, computing predictions (*forward pass)*
+
+-   compare those predictions to the ground truth and quantify the loss
+
+-   run backwards through the network, computing the gradients that indicate how the weights should be changed
+
+-   actually update the weights, making use of the requested learning rate
+
+Here is the template we're going to fill:
+
+```{r}
+for (t in 1:200) {
+    
+    ### -------- Forward pass -------- 
+    
+    # here we'll compute the prediction
+    
+    
+    ### -------- compute loss -------- 
+    
+    # here we'll compute the sum of squared errors
+    
+
+    ### -------- Backpropagation -------- 
+    
+    # here we'll pass through the network, calculating the required gradients
+    
+
+    ### -------- Update weights -------- 
+    
+    # here we'll update the weights, subtracting portion of the gradients 
+}
+```
+
+The forward pass effectuates two affine transformations, one for the hidden and output layers each. In-between, ReLU activation is applied:
+
+```{r}
+  # compute pre-activations of hidden layers (dim: 100 x 32)
+  # torch_mm does matrix multiplication
+  h <- x$mm(w1) + b1
+  
+  # apply activation function (dim: 100 x 32)
+  # torch_clamp cuts off values below/above given thresholds
+  h_relu <- h$clamp(min = 0)
+  
+  # compute output (dim: 100 x 1)
+  y_pred <- h_relu$mm(w2) + b2
+```
+
+Our loss here is mean squared error:
+
+```{r}
+  loss <- as.numeric((y_pred - y)$pow(2)$sum())
+```
+
+Calculating gradients the manual way is a bit tedious, but it can be done:
```

Comment:
Not sure if this is the right wording but I would worry unfamiliar with `torch` might assume you have to calculate gradients by hand. You kinda mention in the closing paragraph that this will get easier in the next post, but maybe worth mentioning here explicitly that `torch` does support automatic differentation?

```suggestion
Calculating gradients the manual way is a bit tedious, but it can be done. Do not worry, `torch` supports automatic differentiation, but for this post we will compute the gradients by hand:
```

Reply:
And i thought i was the official team paranoic ;-))
yeah but i see what you mean, adding a footnote making this clear